### ISSUE TITLE
PermissionError causes unhandled exception

### ISSUE THREAD

**Issue by turekl** [rdiff-backup/rdiff-backup#310]
With rdiff-backup 1.3.3 an unaccessible directory was skipped with just an error reported in log. With rdiff-backup 2.0.0 the same situation causes unhandled TypeError and crash. Unfortunately, excluding the directory does not help due to issue #245. Full output follows:

Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/robust.py", line 35, in check_common_error
    return function(*args)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/rpath.py", line 1410, in append
    return self.__class__(self.conn, self.base, self.index + ext)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/rpath.py", line 1082, in __init__
    self.setdata()
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/rpath.py", line 1108, in setdata
    self.data = self.conn.rpath.make_file_dict(self.path)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/rpath.py", line 386, in make_file_dict
    statblock = os.lstat(filename)
PermissionError: [Errno 13] Permission denied: b'/home/8an/Documents/Praha12.Net/map/www'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/log.py", line 299, in get_indexpath
    return obj.get_safeindexpath()
AttributeError: 'tuple' object has no attribute 'get_safeindexpath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python-exec/python3.6/rdiff-backup", line 32, in <module>
    rdiff_backup.Main.error_check_Main(sys.argv[1:])
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/Main.py", line 390, in error_check_Main
    Main(arglist)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/Main.py", line 412, in Main
    take_action(rps)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/Main.py", line 348, in take_action
    Backup(rps[0], rps[1])
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/Main.py", line 434, in Backup
    backup.Mirror_and_increment(rpin, rpout, incdir)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/backup.py", line 55, in Mirror_and_increment
    DestS.patch_and_increment(dest_rpath, source_diffiter, inc_rpath)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/backup.py", line 279, in patch_and_increment
    for diff in rorpiter.FillInIter(source_diffiter, dest_rpath):
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/rorpiter.py", line 202, in FillInIter
    for rp in rpiter:
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/backup.py", line 114, in get_diffs
    for dest_sig in dest_sigiter:
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/backup.py", line 196, in get_sigs
    for src_rorp, dest_rorp in cls.CCPP:
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/backup.py", line 361, in __next__
    source_rorp, dest_rorp = next(self.iter)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/rorpiter.py", line 95, in Collate2Iters
    relem1 = next(riter1)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/rorpiter.py", line 376, in __next__
    next_elem = next(self.iter)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/selection.py", line 138, in Iterate_fast
    rpath, val = next(diryield_stack[-1])
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/selection.py", line 122, in diryield
    error_handler, rpath.append, (filename, ))
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/robust.py", line 44, in check_common_error
    return error_handler(exc, *args)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/selection.py", line 109, in error_handler
    exc)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/log.py", line 313, in write_if_open
    cls.write(error_type, rp, exc)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/log.py", line 284, in write
    logstr = cls.get_log_string(error_type, rp, exc)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/log.py", line 322, in get_log_string
    return "%s: '%s' %s" % (error_type, cls.get_indexpath(rp), exc)
  File "/usr/lib64/python3.6/site-packages/rdiff_backup/log.py", line 302, in get_indexpath
    return "/".join(obj)
TypeError: sequence item 0: expected str instance, bytes found

**Comment by ericzolf** [comment]
OK, it looks like you stumbled upon quite a log of bugs at once (I see at least 3 of those), we need to improve our test coverage.

**Comment by ericzolf** [comment]
Can you please be more precise on the combination of access rights necessary to reproduce the error? I don't seem to be able to reproduce the right combination to get the error you got. The output of the command `ls -laR /home/8an/Documents/Praha12.Net/map` might be sufficient, unless acls or similar are involved.

**Comment by turekl** [comment]
The directory `/home/8an/Documents/Praha12.Net/map/www` is actually a FUSE sshfs mount, so it wasn't accessible even though I run rdiff-backup as root. As a workaround I remounted it with `-o allow_other` to prevent the crash. Since then rdiff-backup-2.0.0 ran without a hitch every day.

I guess on normal filesystems rdiff-backup checks permissions and skips directories it's not allowed to access, so it never triggers the permission error. Of course I don't expect you to handle my special case, just hope you could make it work as it did in 1.3.3, log the error instead of a crash. Thanks!

**Comment by ericzolf** [comment]
Understood, then I stop trying to reproduce, I think anyway that I have understood and fixed the underlying issue in PR #312 

I would be great if you could test the branch/PR on your setup and tell me if it really fixed the issue.

**Comment by turekl** [comment]
I tested the branch ericzolf-fix-log-tuples-310 on an empty directory with the same SSHFS mount and can confirm the bug is fixed. This is the full output with `--terminal-verbosity=6`:
```
Using rdiff-backup version 2.0.1.dev2+g4a30c1b
        with cpython /usr/lib/python-exec/python3.7/python3 version 3.7.6
        on Linux-4.19.66-gentoo-x86_64-Intel-R-_Xeon-R-_CPU_E3-1226_v3_@_3.30GHz-with-gentoo-2.6, fs encoding utf-8
Unable to import win32security module. Windows ACLs
not supported by filesystem at src
Exception '[Errno 13] Permission denied: b'src/mount'' raised of class '<class 'PermissionError'>':
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/robust.py", line 35, in check_common_error
    return function(*args)
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/rpath.py", line 1410, in append
    return self.__class__(self.conn, self.base, self.index + ext)
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/rpath.py", line 1082, in __init__
    self.setdata()
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/rpath.py", line 1108, in setdata
    self.data = self.conn.rpath.make_file_dict(self.path)
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/rpath.py", line 386, in make_file_dict
    statblock = os.lstat(filename)

ListError: '(b'mount',)' [Errno 13] Permission denied: b'src/mount'
-----------------------------------------------------------------
Detected abilities for source (read only) file system:
  Access control lists                         On
  Extended attributes                          On
  Windows access control lists                 Off
  Case sensitivity                             Off
  Escape DOS devices                           Off
  Escape trailing spaces                       Off
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
Making directory dst/rdiff-backup-data/rdiff-backup.tmp.0
Making directory dst/rdiff-backup-data/rdiff-backup.tmp.0/hl
Hard linking dst/rdiff-backup-data/rdiff-backup.tmp.0/hl/hardlinked_file2 to dst/rdiff-backup-data/rdiff-backup.tmp.0/hardlinked_file1
Unable to import win32security module. Windows ACLs
not supported by filesystem at dst/rdiff-backup-data/rdiff-backup.tmp.0
Removing directory dst/rdiff-backup-data/rdiff-backup.tmp.0
-----------------------------------------------------------------
Detected abilities for destination (read/write) file system:
  Ownership changing                           On
  Hard linking                                 On
  fsync() directories                          On
  Directory inc permissions                    On
  High-bit permissions                         On
  Symlink permissions                          Off
  Extended filenames                           On
  Windows reserved filenames                   Off
  Access control lists                         On
  Extended attributes                          On
  Windows access control lists                 Off
  Case sensitivity                             On
  Escape DOS devices                           Off
  Escape trailing spaces                       Off
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
Backup: escape_dos_devices = 0
Backup: escape_trailing_spaces = 0
Writing mirror marker dst/rdiff-backup-data/current_mirror.2020-03-28T16:17:46+01:00.data
Starting increment operation src to dst
Exception '[Errno 13] Permission denied: b'src/mount'' raised of class '<class 'PermissionError'>':
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/robust.py", line 35, in check_common_error
    return function(*args)
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/rpath.py", line 1410, in append
    return self.__class__(self.conn, self.base, self.index + ext)
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/rpath.py", line 1082, in __init__
    self.setdata()
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/rpath.py", line 1108, in setdata
    self.data = self.conn.rpath.make_file_dict(self.path)
  File "/home/8an/Documents/Temp/rdiff-backup/lib/python3.7/site-packages/rdiff_backup-2.0.1.dev2+g4a30c1b-py3.7-linux-x86_64.egg/rdiff_backup/rpath.py", line 386, in make_file_dict
    statblock = os.lstat(filename)

ListError: '(b'mount',)' [Errno 13] Permission denied: b'src/mount'
Writing mirror_metadata diff
Cleaning up
```
